[XLA:CPU][oneDNN] Prevent F32 fallback for BF16 rewritable dots

akhilgoe opened the pull request:
This PR prevents some oneDNN rewritable BF16 dots from being converted to F32. In particular:
1. This PR prevents the oneDNN rewriter from performing layout checks before the layout assignment passes, thus avoiding the upcasting of dots with column-major layouts that might later be converted to row-major layouts.
2. In case the BF16 dots are not rewritable to oneDNN custom calls, this PR upcasts the BF16 dots to F32.
3. Currently, the oneDNN rewriter bypasses the layout checks for batch dots. This PR also introduces the layout checks for batch dots during the rewrite process.

NaiyerRizz commented:
Hi @akhilgoe 
Can you please see as few checks are failing.
Thank you

penpornk commented on xla/service/cpu/onednn_contraction_rewriter.cc:
```diff
@@ -400,10 +400,11 @@ bool OneDnnContractionRewriter::ShouldRewriteDot(
-      (!IsRowMajor(lhs_shape) || !IsRowMajor(rhs_shape) ||
-       !IsRowMajor(output_shape))) {
-    return false;
+  // Skip row-major layout check before layout-pass
```
Nit:
```suggestion
  // Skip row-major layout check before layout-assignment pass
```

penpornk commented on xla/service/cpu/onednn_contraction_rewriter.cc:
```diff
@@ -400,10 +400,11 @@ bool OneDnnContractionRewriter::ShouldRewriteDot(
   // scenarios in last two dimensions.
   // Col-major layouts are corrected to row-majow for BatchDot operation as
   // part of the layout-pass.
```
Nit:
```suggestion
  // Col-major layouts are corrected to row-major for BatchDot operation as
  // part of the layout-assignment pass.
```

penpornk commented on xla/service/cpu/onednn_contraction_rewriter.cc:
```diff
@@ -1015,6 +1018,34 @@ class OneDnnContractionRewriteVisitor : public DfsHloRewriteVisitor {
   }
+
+  // This function upcasts BF16 dots to F32 if we are unable to rewrite them to
+  // onednn custom calls.
```
Nit: 
```suggestion
  // oneDNN custom calls.
```

penpornk reviewed (changes requested):
Thank you for the PR. Could you please add a test case for this?

akhilgoe commented:
Thanks @NaiyerRizz for notifying me about the failing tests. I looked at the logs and I think the error might have been caused due to a missing rocm repository. I'll take a look again if the tests still fail.

akhilgoe commented:
Thanks @penpornk for the review! I think I have addressed all your comments. I also added a test that would fail without the changes in this PR. Please take a look.

penpornk commented on xla/service/cpu/tests/onednn_matmul_test.cc:
```diff
@@ -1501,6 +1501,27 @@ TEST_F(MatmulTest, WeightsPrepackAndScratch) {
+
+  EXPECT_TRUE(RunAndCompare(matmul_module_str, ErrorSpec(1e-2, 1e-2)));
+  MatchOptimizedHlo(matmul_module_str,
+                    R"(
+  ; CHECK:     (bf16[500,500]{1,0}, u8[{{.*}}]{0}) custom-call(%{{[a-z,A-Z,0-9,\.]*}}, %{{[a-z,A-Z,0-9,\.]*}}), custom_call_target="__onednn$matmul",
+  )");
+}
```
Nit: Maybe split the checks to multiple lines using CHECK and CHECK-SAME? See lines 1494-1500 for examples.
E.g.,
```
; CHECK: (bf16[500,500]{1,0}, u8[{{.*}}]{0})
; CHECK-SAME: custom_call_target="__onednn$matmul"
```

penpornk commented on xla/service/cpu/tests/onednn_matmul_test.cc:
```diff
@@ -1501,6 +1501,27 @@ TEST_F(MatmulTest, WeightsPrepackAndScratch) {
+    arg.0 = bf16[500,500]{0,1} parameter(0)
+    arg.1 = bf16[500,500]{1,0} parameter(1)
+    transpose.0 = bf16[500,500]{0,1} transpose(arg.1), dimensions={1,0}
+    ROOT dot.0 = bf16[500,500]{1,0} dot(arg.0, arg.1), lhs_contracting_dims={1}, rhs_contracting_dims={0}
```
Nit: Can we wrap this line so it's <= 80 characters long? Can simply do:
```suggestion
    ROOT dot.0 = bf16[500,500]{1,0} dot(arg.0, arg.1), lhs_contracting_dims={1},
      rhs_contracting_dims={0}
```

penpornk reviewed (changes requested):
Thank you for the quick response! I have two more nits please.

akhilgoe commented:
Thanks @penpornk! I have created a commit with your suggested changes.

akhilgoe commented:
Yes, I tested before pushing the commit.
